sphinx/directives/code.py

```diff
@@ -190,202 +372,53 @@ class LiteralInclude(Directive):
-        if linespec:
-            try:
-                hl_lines = [x + 1 for x in parselinenos(linespec, len(lines))]
-            except ValueError as err:
-                return [document.reporter.warning(str(err), line=self.lineno)]
-        else:
-            hl_lines = None
-
-        start_str = self.options.get('start-after')
-        start_inclusive = False
-        if self.options.get('start-at') is not None:
-            start_str = self.options.get('start-at')
-            start_inclusive = True
-        end_str = self.options.get('end-before')
-        end_inclusive = False
-        if self.options.get('end-at') is not None:
-            end_str = self.options.get('end-at')
-            end_inclusive = True
-        if start_str is not None or end_str is not None:
-            use = not start_str
-            res = []
-            for line_number, line in enumerate(lines):
-                if not use and start_str and start_str in line:
-                    if 'lineno-match' in self.options:
-                        linenostart += line_number + 1
-                    use = True
-                    if start_inclusive:
-                        res.append(line)
-                elif use and end_str and end_str in line:
-                    if end_inclusive:
-                        res.append(line)
-                    break
-                elif use:
-                    res.append(line)
-            lines = res
-
-        prepend = self.options.get('prepend')
-        if prepend:
-            lines.insert(0, prepend + '\n')
-
-        append = self.options.get('append')
-        if append:
-            lines.append(append + '\n')
-
-        text = ''.join(lines)
-        if self.options.get('tab-width'):
-            text = text.expandtabs(self.options['tab-width'])
-        retnode = nodes.literal_block(text, text, source=filename)
-        set_source_info(self, retnode)
-        if diffsource:  # if diff is set, set udiff
-            retnode['language'] = 'udiff'
-        if 'language' in self.options:
-            retnode['language'] = self.options['language']
-        retnode['linenos'] = 'linenos' in self.options or \
-                             'lineno-start' in self.options or \
-                             'lineno-match' in self.options
-        retnode['classes'] += self.options.get('class', [])
-        extra_args = retnode['highlight_args'] = {}
-        if hl_lines is not None:
-            extra_args['hl_lines'] = hl_lines
-        extra_args['linenostart'] = linenostart
-        env.note_dependency(rel_filename)
-
-        caption = self.options.get('caption')
-        if caption is not None:
-            if not caption:
-                caption = self.arguments[0]
-            try:
+        try:
+            rel_filename, filename = env.relfn2path(self.arguments[0])
+            env.note_dependency(rel_filename)
+
+            reader = LiteralIncludeReader(filename, self.options, env.config)
```

Comment:
Now `LiteralIncludeReader` is a core implementation of the `literalinclude` directive.
From now on, the responsibility of `LiteralInclude.run()` is converting the result of `LiteralIncludeReader` to the doctree and handles errors.